server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,456 @@
+
+    MirroringMigrationService(ProjectManager projectManager, CommandExecutor commandExecutor) {
+        this.projectManager = projectManager;
+        this.commandExecutor = commandExecutor;
+    }
+
+    void migrate() throws Exception {
+        if (hasMigrationLog()) {
+            logger.debug("Mirrors and credentials have already been migrated. Skipping auto migration...");
+            return;
+        }
+
+        // Enter read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.REPLICATION_ONLY))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Starting Mirrors and credentials migration ...");
+        if (commandExecutor instanceof ZooKeeperCommandExecutor) {
+            logger.debug("Waiting for 30 seconds to make sure that all cluster have been notified of the " +
+                         "read-only mode ...");
+            Thread.sleep(30000);
+        }
+
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        try {
+            for (Project project : projectManager.list().values()) {
+                logger.info("Migrating mirrors and credentials in the project: {} ...", project.name());
+                boolean processed = false;
+                final MetaRepository repository = project.metaRepo();
+                processed |= migrateCredentials(repository);
+                // Update the credential IDs in the mirrors.json file.
+                processed |= migrateMirrors(repository);
+                if (processed) {
+                    logger.info("Mirrors and credentials in the project: {} have been migrated.",
+                                project.name());
+                } else {
+                    logger.info("No legacy configurations of mirrors and credentials found in the project: {}.",
+                                project.name());
+                }
+            }
+            logMigrationJob();
+        } catch (Exception ex) {
+            final MirrorMigrationException mirrorException = new MirrorMigrationException(
+                    "Failed to migrate mirrors and credentials. Rollback to the legacy configurations", ex);
+            try {
+                rollbackMigration();
+            } catch (Exception ex0) {
+                ex0.addSuppressed(mirrorException);
+                throw new MirrorMigrationException("Failed to rollback the mirror migration:", ex0);
+            }
+            throw mirrorException;
+        }
+
+        // Exit read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.WRITABLE))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Mirrors and credentials migration has been completed. (took: {} ms.)",
+                    stopwatch.elapsed().toMillis());
+
+        shortWords = null;
+    }
+
+    private void logMigrationJob() throws Exception {
+        final ImmutableMap<String, Instant> data = ImmutableMap.of("timestamp", Instant.now());
```

Comment:
How about also adding the total number of the projects that have been migrated so we can easily know if the migration is successfully done?